shared/middleware/helpers/storage_apps.rb

```diff
@@ -140,24 +140,43 @@ def get_abuse(channel_id)
 
   # Determine if the current user can view the project
   def get_sharing_disabled(channel_id, current_user_id)
-    user_storage_owner_id, _ = storage_decrypt_channel_id(channel_id)
+    user_storage_owner_id, storage_app_id = storage_decrypt_channel_id(channel_id)
     owner_id = user_storage_ids_table.where(id: user_storage_owner_id).first[:user_id]
 
     # Sharing of a project is not disabled for the project owner
     # or the teachers of the project owner
+    # or if the current user paired with the owner
     if current_user_id == owner_id
       return false
     elsif teaches_student?(owner_id, current_user_id)
       return false
+    elsif get_user_sharing_disabled(owner_id)
+      !users_paired_on_level?(storage_app_id, current_user_id, owner_id, user_storage_owner_id)
     else
-      return get_user_sharing_disabled(owner_id)
+      return false
     end
 
   # Default to sharing disabled if there is an error
   rescue ArgumentError, OpenSSL::Cipher::CipherError
     true
   end
 
+  def users_paired_on_level?(storage_app_id, current_user_id, owner_id, user_storage_owner_id)
+    channel_tokens_table = DASHBOARD_DB[:channel_tokens]
+    level_id = channel_tokens_table.select(:level_id).where(storage_app_id: storage_app_id, storage_id: user_storage_owner_id)
```

Comment:
something tricky may be happening here -- `where` on this line should be giving you back an array-like structure, which doesn't cause an error later because `where` on later lines can accept arrays in addition to individual values. 

I'd recommend changing `where` to `first` on this line, so that level_id is a single value (or nil).

Reply:
my recommended fix here is not quite right, because it leaves you with an object `{level_id: value}` instead of just the value. Here are a couple of examples that could be better to follow:

https://github.com/code-dot-org/code-dot-org/blob/9d1911a4e593c97997bc86a92baff1786d755ae6/shared/middleware/helpers/storage_id.rb#L92-L100